ethereum/eip712/preprocess.go

```diff
@@ -0,0 +1,79 @@
+package eip712
+
+import (
+	"fmt"
+
+	"github.com/cosmos/cosmos-sdk/client"
+	codectypes "github.com/cosmos/cosmos-sdk/codec/types"
+	cosmoskr "github.com/cosmos/cosmos-sdk/crypto/keyring"
+	"github.com/cosmos/cosmos-sdk/types/tx/signing"
+	authtx "github.com/cosmos/cosmos-sdk/x/auth/tx"
+	"github.com/evmos/ethermint/types"
+)
+
+// PreprocessLedgerTx reformats Ledger-signed Cosmos transactions to match the fork expected by Ethermint
+// by including the signature in a Web3Tx extension and sending a blank signature in the body.
+func PreprocessLedgerTx(chainID string, keyType cosmoskr.KeyType, txBuilder client.TxBuilder) error {
+	// Only process Ledger transactions
+	if keyType != cosmoskr.TypeLedger {
+		return nil
+	}
+
+	// Init extension builder to set Web3 extension
+	extensionBuilder, ok := txBuilder.(authtx.ExtensionOptionsTxBuilder)
+	if !ok {
+		return fmt.Errorf("cannot cast TxBuilder to ExtensionOptionsTxBuilder")
+	}
+
+	// Get signatures from TxBuilder
+	sigs, err := txBuilder.GetTx().GetSignaturesV2()
+	if err != nil {
+		return fmt.Errorf("could not get signatures: %w", err)
+	}
+
+	// Verify single-signer
+	if len(sigs) != 1 {
+		return fmt.Errorf("invalid number of signatures, expected 1 and got %v", len(sigs))
+	}
+
+	signature := sigs[0]
+	sigBytes := signature.Data.(*signing.SingleSignatureData).Signature
```

Comment:
Is the `signature.Data` created by signing the EIP712 representation of this transaction? Because we can not just move the signature of a normal transaction to the extension, because when validating the signature it will fail.

Reply:
Yes, this is verified by checking that the keytype is of type Ledger:
```
if keyType != cosmoskr.TypeLedger {
    return nil
}
```
Since the only way to sign Cosmos transactions with Ledger is via EIP-712, this should enforce the condition.